# **データマイニングと情報可視化: 演習**
# **Data mining and information visualization: Exercise**
<font size="4">
第3回　2023年12月19日(火)<br>
データマイニング手法1-2：マーケットバスケット分析（データ解析と可視化）<br>
Method 1-2: Market basket analysis (Analyze data &amp;visualization)<br>
</font>

[レポート / Assignment](Rep_Week3.ipynb) <span style="color: black;font-weight:bold;" >締め切り 2023年12月22日(金) 23:59 JST</span>

<img src="img/w3_marketbasket.png" width="500px">

In [22]:
# まずはライブラリをインポート / Import the libraries.
import numpy as np
import numpy.random as random
import scipy as sp
from pandas import Series, DataFrame
import pandas as pd

# 可視化ライブラリ / Visualization libraries
import matplotlib.pyplot as plt
import matplotlib as mpl
import seaborn as sns
%matplotlib inline

# 機械学習ライブラリ (念のため)/ Machine learning library (may not be used, but just in case..)
import sklearn

# 小数第3位まで表示 / display up to the 3rd decimal place
%precision 3

'%.3f'

### Useful shortcuts

<b>セルの実行 / Run selected cells and don't advance</b><br>
Ctrl-Enter

<b>セルの実行して下のセルに移動 / Run selected cells and advance</b><br>
Shift-Enter

<b>複数行のコメントアウト / Comment out multiple lines</b><br>
Ctrl + /<br>
<br>
<b>セルを下に追加 / Add a cell below</b><br>
Esc + b<br>
<br>
<b>セルをマークダウンモード/ Set a cell to Markdown mode</b><br>
Esc + m<br>
<br>
<b>セルを削除/ Delete a cell</b><br>
Esc + d d<br>

## Pandasの基礎 / Basics of Pandas
[参考サイト：Pandas 基礎とよく使う機能メモ](https://qiita.com/ryo111/items/144446bedb9aa75b0acc)　コンパクトにまとまってる。<br>
[参考サイト：pandas 入門 - pandasによるデータ処理の基礎](https://www.yunabe.jp/docs/pandas_basics.html)　説明が丁寧。<br>

In [23]:
# データはこちら。/ Here is the test data.
Nottinghill = pd.DataFrame([[56, 44, 89, "F"],
                  [88, 57, 94, "F"],
                  [74, 88, 46, "M"],
                  [98, 88, 75, "M"],
                  [77, 83, 79, "M"],
                  [85, 67, 75, "F"]],
                  index=["Anna", "Bella", "William", "Bernie", "Spike", "Honey"],
                  columns=["History", "Literature", "Philosophy","sex"])
Nottinghill

History  Literature  Philosophy sex
Anna          56          44          89   F
Bella         88          57          94   F
William       74          88          46   M
Bernie        98          88          75   M
Spike         77          83          79   M
Honey         85          67          75   F

## DataFrame 'Nottinghill' の行数と列数を数えて表示してください。
## Please display the numbers of rows and columns of a DataFrame, 'Nottinghill'.

[Week2: shape](Ex_Week2.ipynb#shape)

In [24]:
# # # Number of columns
print(len(Nottinghill))

# # # Number of rows
print(len(Nottinghill.columns))


6
4


## 平均値 / Average

In [25]:
# History の平均値 / Average of History
History_mean = Nottinghill['History'].mean()
print(History_mean)


79.66666666666667


In [26]:
# Literature の平均値 / Average of Literature
History_mean = Nottinghill['Literature'].mean()
print(History_mean)

# Philosophy の平均値 / Average of Philosophy
History_mean = Nottinghill['Philosophy'].mean()
print(History_mean)


71.16666666666667
76.33333333333333


## Groupby でグループごとに平均値を出す / Calculate average for each group with groupby
[Week2: groupby](Ex_Week2.ipynb#groupby)

In [27]:
# GROUPBY を使いましょう。
# Please use GROUPBY.
Nottinghill.groupby


<bound method DataFrame.groupby of          History  Literature  Philosophy sex
Anna          56          44          89   F
Bella         88          57          94   F
William       74          88          46   M
Bernie        98          88          75   M
Spike         77          83          79   M
Honey         85          67          75   F>

In [28]:
# 女性("F")と男性("M")のそれぞれについて、History の平均値を計算してください。
# Please calculate average score of History for female ("F") and male ("M").
NG = Nottinghill.groupby('sex')

# get_groupを使って、Female のデータだけ抽出 / Extract data of females with get_group.
NG_F = NG.get_group('F')

# get_groupを使って、Male のデータだけ抽出 / Extract data of males with get_group.
NG_M = NG.get_group('M')

# Average of History of female
NG_F_history = NG_F['History'].mean()
print(NG_F_history)

# Average of History of male
NG_M_history = NG_M['History'].mean()
print(NG_M_history)

76.33333333333333
83.0


In [29]:
# 一度に全部やるなら / If you want to calculate at once...
Nottinghill.groupby('sex').mean().History

Nottinghill.groupby('sex').get_group('F').mean(numeric_only=True).History

76.333

## グラフをプロットする練習 / Practice to draw a graph
History を横軸、Literature を縦軸にして、散布図を書いてみてください。<br>
Please draw a scattering plot with History and Literature. The horizontal and vertical axes would be History and Literature, respectively.
<br><br>
[参考サイト：matplotlibでよく使われるやつを改めてまとめた](https://qiita.com/kakiuchis/items/798c00f54c9151ab2e8b)<br>
[参考サイト：matplotlib で散布図 (Scatter plot) を描く](https://pythondatascience.plavox.info/matplotlib/%E6%95%A3%E5%B8%83%E5%9B%B3)<br>
[参考サイト：Python / matplotlib を利用した散布図の描き方](https://stats.biopapyrus.jp/python/scatter.html)<br>
[EN: Matplotlib Documentation](https://matplotlib.org/)


In [30]:
# Plot data with scattering plot
# plt.scatter(NF.History,NF.Literature,s=200,c='pink',marker='*',alpha=0.5,linewidths='2',edgecolors='red',label="Female")
# plt.scatter(NM.History,NM.Literature,s=200,c='blue',marker='*',alpha=0.5,linewidths='2',edgecolors='blue',label="Male")
plt.scatter(NF.History,NF.Literature,label="Female",s=100,c='red',marker='^',alpha=0.5)
plt.scatter(NM.History,NM.Literature,label="Male",s=100,c='blue',marker='o',alpha=0.5)
# plt.title('History vs. Literature')
plt.ylabel('Literature')
plt.xlabel('History')
plt.legend(bbox_to_anchor=(1.4, 1), loc='upper right', borderaxespad=0, fontsize=18) #凡例
plt.show()


NameError: name 'NF' is not defined

## 集合の演算 / Set operations
[Week 2: intersection](Ex_Week2.ipynb#setinsec)

In [ ]:
# Casts in the films...
# ノッティングヒルの恋人　Notting Hill (1999)
Nottinghill = {'Julia Roberts', 'Hugh Grant', 'Hugh Bonnevill', 'Gina McKee '}

# プリティウーマン　Pretty Woman (1990)
PrettyWoman = {'Julia Roberts', 'Richard Gere', 'Hector Elizondo'}

# プリティブライド　Runaway Bride (1999)
RunawayBride = {'Richard Gere', 'Julia Roberts', 'Joan Cusack','Christopher Meloni', 'Hector Elizondo'}

# トップガン　Top Gun (1986)
TopGun = {'Tom Cruise', 'Val Kilmer', 'Tim Robbins', 'Meg Ryan'}

# めぐり逢えたら　Sleepless in Seattle (1993)
SleeplessinSeattle = {'Meg Ryan', 'Tom Hanks', 'Valerie Wright', 'Bill Pullman'}


In [ ]:
# ノッティングヒルの恋人とプリティウーマンのいずれの映画にも出演していた俳優・女優を見つけて表示してください。
# Please find actors who acted both in Notting hill and Pretty Woman. Then display the result.
intersection_NP = Nottinghill.intersection(PrettyWoman)
print(intersection_NP)


{'Julia Roberts'}


In [ ]:
# プリティウーマンとプリティブライドのいずれの映画にも出演していた俳優・女優を見つけて表示してください。
# Please find actors who acted both in Pretty Woman and Runaway Bride. Then display the result.
intersection_PR = PrettyWoman.intersection(RunawayBride)
print(intersection_PR)

{'Julia Roberts', 'Hector Elizondo', 'Richard Gere'}


In [ ]:
# トップガンと巡り合えたらのいずれの映画にも出演していた俳優・女優を見つけて表示してください。
# Please find actors who acted both in Top Gun and Sleepless in Seattle. Then display the result.
intersection_TS = TopGun.intersection(SleeplessinSeattle)
print(intersection_TS)

{'Meg Ryan'}


<a id="load_excel"></a>

## サンプルデータの読み込み / Load a sample data
購買履歴データを使って、マーケットバスケット分析を具体的に説明していきます。<br>
扱うデータは、`w3_OnlineRetail.xlsx`というファイルです。`data`フォルダに配置してください。<br>

`trans`という変数名で読み込むことにします。`head`を使って先頭の5レコードを表示してみましょう。<br>
<br>
Let's start Market basket analysis with sample purchasing data.<br>
Here, we use a sample data, `w3_OnlineRetail.xlsx`. Please save this file in a folder `data`.<br>

Let's load this file as a variable named trans.<br>
Display the first 5 lines of the file with `head` command. <br>


サンプルデータは、[カリフォルニア大学アーバイン校が提供するオンラインショッピングサイトのトランザクションを含むデータセット](https://archive.ics.uci.edu/ml/datasets/online+retail)を使っています。データ詳細はこちら。<br>
* InvoiceNo: 請求書番号
* StockCode: 商品番号
* Description: 商品詳細
* Quantity: 購入数
* InvoiceDate: 請求書発行日
* UnitPrice: 商品単価
* CustomerID: 顧客番号
* Country: 国


In [32]:
# サンプルファイルを読み込みます。ちょっと時間がかかりますから、焦らず待ちましょう。
# Load the sample file. It takes a while to load the entire file. Please be patient!!
trans = pd.read_excel('data/w3_retail.xlsx')

In [ ]:
## read_excel が使えない場合はこちらを試してください。
# If excel file cannot be loaded with 'pd.read_excel', please use openpyxl.
# If you haven't installed it yet, you need to install it.

# import openpyxl
# from openpyxl import Workbook, load_workbook

## https://www.python-izm.com/third_party/excel/openpyxl/openpyxl_read/
# trans = load_workbook('data/w3_retail.xlsx')

In [ ]:
# 最初の5行を表示してください。
# Display the first 5 lines of the file with `tail` command. 
# ****** に適切な関数を入れてください。/ Fill in the part ***** with an appropriate function.
# Hint: trans.****()
trans.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [ ]:
# データのサイズをチェックしてみよう。****** に適切な関数を入れてください。 / Check the data size. Fill in the part ***** with an appropriate function.
# Hint: trans.*****
trans.shape

(541909, 8)

<a id="data_processing"></a>

### データの整理と確認 / Data processing and check a data file
この購買履歴データにおいて請求書番号を示す`InvoiceNo`の先頭の1文字は、そのトランザクションの状態を示しています。「5」が通常のデータ、「C」がキャンセル、「A」が不明なデータです。まずは次のように`InvoiceNo`の先頭1文字を別の`cancel_flg`という変数として追加します。<br>
<br>
In this purchasing data, the first letter of `IncoiceNo` indicates the transaction state as '5' is normal data, 'C' is cancelled, 'A' is unknown.<br>
Add a variable `cancel_flg` by extracting the first letter of `IncoiceNo`.

<a id="map_lambda"></a>

In [ ]:
# InoivceNoの先頭1文字をcancel_flgとして追加 / Add a variable `cancel_flg` by extracting the first letter of InvoiceNo.
trans['cancel_flg'] = [str(i)[0] for i in trans.InvoiceNo]

# invoicelist = trans.InvoiceNo
# flglist = list()
# for i in invoicelist:
#     flglist.append(str(i)[0])
# trans['cancel_flg'] = flglist

追加したら、それぞれの`cancel_flg`ごとのレコード数を数えて集計します。実務において集計条件は、分析目的とデータマネジメント状態に大きく依存するので、十分に確認するようにしましょう。

In [ ]:
# trans['cancel_flg'].head()
trans.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country cancel_flg  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom          5  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom          5  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom          5  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom          5  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom          5

In [ ]:
# cancel_flgでグルーピングして集計 / Make a group of each transaction state.
trans_group = trans.groupby('cancel_flg')

# 各トランザクション状態の取引は、それぞれいくつずつありましたか?? / How many transactions of each state?
# Normal data (5)
normaldata = trans_group.get_group('5')
print(len(normaldata))

# Cancelled (C)
Canncelled = trans_group.get_group('C')
print(len(Canncelled))

# Unknown (A)
Unknown = trans_group.get_group('A')
print(len(Unknown))

532618
9288
3


## 欠損値があるか調べてみましょう。

[pandasで欠損値NaNが含まれているか判定、個数をカウント](https://note.nkmk.me/python-pandas-nan-judge-count/)<br>
[Ex_Week0_Pandas.ipynb: isnull](Ex_Week0_Pandas.ipynb#countnan)

In [ ]:
print(trans.isnull().sum())

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
cancel_flg          0
dtype: int64


## Description と　CustomerID　に欠損値がありました。

以下、通常のデータである「5」であり、かつ、`CustomerID`が欠損していないデータだけを扱うことにします。これらのデータに絞り込むため、次のようにします。<br>
<br>
From now on, we are going to analyze only the normal data which has `CustomerID` as follows.

In [ ]:
# extract data only when cancel_flg is 5, and CustomerID is not null.
trans = trans[(trans.cancel_flg == '5') & (trans.CustomerID.notnull())]

In [ ]:
# 有効なデータのサイズをチェックして、もともとデータサイズと比べてみよう。****** に適切な関数を入れてください。
# Check the size of the available data. Compare it with the original data size. Fill in the part ***** with an appropriate function.
# Hint: trans.*****
trans.info()

<class 'pandas.core.frame.DataFrame'>
Index: 397924 entries, 0 to 541908
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    397924 non-null  object        
 1   StockCode    397924 non-null  object        
 2   Description  397924 non-null  object        
 3   Quantity     397924 non-null  int64         
 4   InvoiceDate  397924 non-null  datetime64[ns]
 5   UnitPrice    397924 non-null  float64       
 6   CustomerID   397924 non-null  float64       
 7   Country      397924 non-null  object        
 8   cancel_flg   397924 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(5)
memory usage: 30.4+ MB


データ行の数が541909から397924に減っています。

## アソシエーションルール / Association rule

データの準備ができたところで、アソシエーションルールについて説明します。
まずは、購買回数トップ5の製品番号を確認しておきましょう。製品番号は`stockCode`列に格納されています。

Pandasの`Series`オブジェクトが持つ`value_counts`メソッドを使うと、それぞれの内容別のレコード件数がデフォルトでは降順で得られます。そこで`head`を使って、上位5件を表示すると、次のようになります。

<a id="value_count"></a>

### ここで使う関数: value_count()
### Required functions: : value_count()
value_count(): 要素の出現頻度を数え、出現頻度の高い順にPandas.Seriesで返します。<br>
[value_counts() @ note.nkmk.me](https://note.nkmk.me/python-pandas-value-counts/)<br>
 value_count(): It counts the number of appearance of an element, and return the results in a descending order as Pandas.Series.<br>
[value_counts() @ pandas.pydata.org](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.value_counts.html)

In [ ]:
# 商品コード(StockCode)ごとに件数を数え、上位5件を表示 / Display the number of appearance of the item number in a descending order.
trans['StockCode'].value_counts().head()

StockCode
85123A    2035
22423     1724
85099B    1618
84879     1408
47566     1397
Name: count, dtype: int64

## Description からどんな商品が売れているか確認する。

<img src="img/w3_products.png" width="500px">

一番売れたのは吊るすタイプのろうそく台、2番目はケーキスタンドでした。

In [ ]:
# Description に商品詳細が書いてある。
# Stock codeが数字だけだとクォーテーションマークはいらない。Stock codeにアルファベットが含まれると、それ全体が文字列として認識されるのでクォーテーションマークが必要になる。
# Description contains the details of the item.
# If the Stock code is only numbers, no quotation marks are needed; if the Stock code contains alphabetic characters, the whole code is recognized as a string, so quotation marks are needed.

basket_85123A = (trans[trans['StockCode']=="85123A"])
basket_85123A.head()

basket_22423 = (trans[trans['StockCode']==22423])
basket_22423.head()

InvoiceNo StockCode               Description  Quantity  \
880     536477     22423  REGENCY CAKESTAND 3 TIER        16   
936     536502     22423  REGENCY CAKESTAND 3 TIER         2   
1092    536525     22423  REGENCY CAKESTAND 3 TIER         2   
1155    536528     22423  REGENCY CAKESTAND 3 TIER         1   
1197    536530     22423  REGENCY CAKESTAND 3 TIER         1   

             InvoiceDate  UnitPrice  CustomerID         Country cancel_flg  
880  2010-12-01 12:27:00      10.95     16210.0  United Kingdom          5  
936  2010-12-01 12:36:00      12.75     16552.0  United Kingdom          5  
1092 2010-12-01 12:54:00      12.75     14078.0  United Kingdom          5  
1155 2010-12-01 13:17:00      12.75     15525.0  United Kingdom          5  
1197 2010-12-01 13:21:00      12.75     17905.0  United Kingdom          5

以下では上記のトップ5の商品のうち、全体で第1位の「85123A」と第3位の「85099B」に関して、アソシエーションルールとその支持度、確信度、リフトを調べます。<br>
<br>
Now we are going to investigate the association rule (support, confidence, and lift) between the items with the item number number "85123A" and "85099B".<br>
The item "85123A" was sold the most, and the item "85099B" was sold the third most.

<a id="support"></a>

### 支持度 (support)
<img src="img/w3_support.png" width="400px"><br>
アソシエーションルールの支持度とは、ある商品（ここでは85123A）と別の商品（85099B）が併売されたバスケットの数（`InvoiceNo`の数）が全体の購入に占める割合です。<br>
以下で、商品85123Aを購入するなら商品85099Bする、というアソシエーションルールの支持度を計算してみます。<br>
<br>
Support is defined as the number of baskets (or the ratio of the number of baskets to the total number of baskets) in which "85123A" and "85099B" were sold at the same time. <br>
Let's calculate the support!!

まず、トランザクションデータ（すべての購入データ）に登場するバスケットの数（`InvoiceNo`の数）をカウントします。<br>
最初に、すべての`InvoiceNo`を`trans_all`として抽出します。<br>
集合型とすることで、`InvoiceNo`を重複のない状態で保持できます。<br>
集合型(set) については、[こちら](https://www.sejuku.net/blog/21923)。<br>
[Week 2: set](Ex_Week2.ipynb#set)
<br>
First, calculate the number of baskets (number of `InvoiceNo`) in the entire purchasing data.<br>
Instead of a list, we use a set, which is helpful to avoid overlap of the data.<br>
For more details on sets, check [the reference](https://docs.python.org/2/library/sets.html).<br>
[Week 2: set](Ex_Week2.ipynb#set)<br>

In [33]:
# すべてのInvoiceNoをtrans_allとして集合型を定義 / Define a set of all of the `InvoiceNo` as `trans_all`.
trans_all = set(trans.InvoiceNo) 
print(len(trans_all))

25900


In [34]:
trans.InvoiceNo.head()

0    536365
1    536365
2    536365
3    536365
4    536365
Name: InvoiceNo, dtype: object

item A, '85123A', と item B, '85099B', を含む`InvoiceNo`を抽出する（下記では`trans_a`ならびに`trans_b`）。<br>
'StockCode' は商品番号。<br>
<br>
Extract the baskets which contains item A and item B as `trans_a` and `trans_b` respectively, where item A is '85123A' and item B is '85099B'.<br>
'StockCode' denotes the item number.

In [35]:
# 商品85123Aを購入したバスケット(trans_a)の集合を作る / The basket which contains item A is extracted as trans_a.
trans_a = set(trans[trans['StockCode']=='85123A'].InvoiceNo)
print(len(trans_a))

# 商品85099Bを購入したバスケット(trans_b)の集合を作る / The basket which contains item B is extracted as trans_b.


2246


次に両商品を含むバスケットを`trans_ab`として抽出します。`trans_a` と `trans_b`の積集合をとります。<br>
Extract the baskets which contains both item A and item B by taking product set of `trans_a` and `trans_b`.

In [ ]:
# 商品85123Aおよび85099Bを購入したデータをtrans_abとする / Defined the basket which contains both item A and B as trans_ab.
trans_ab = 


ルールの支持度は、ルールに含まれる両商品を含むバスケットの数が全体の購入に占める割合です。<br>

Support is defined as the ratio of the number of baskets to the total number of baskets in which item A and B were sold at the same time. <br>


In [36]:
# trans_ab の、両商品を含むバスケットの数を表示 / Display the number of baskets which contains both item A and B.
# A and B = len(trans_ab)
print('両商品を含むバスケットの数:{}'.format(len(trans_ab)))
print('The number of baskets which contains both item A and B:{}'.format(len(trans_ab)))

NameError: name 'trans_ab' is not defined

<a id="format"></a>

### 関数: .format
### Functions: : .format
`.format` は、文字列の中に数字を埋め込んでprint文で表示させるのに使います。小数点以下何桁を表示するか、も指定できます。<br>
You can embed a value in a string with print command by `.format`. You can also specify which decimal place to be displayed.

In [ ]:
myvar = 2018
print("This is the {}th year.".format(myvar))

myvar = 0.573668
print("The probability to win this game is estimated to be {:.3f}.".format(myvar))

In [ ]:
# support = (A and B) / N
# A and B = len(trans_ab)
# N = len(trans_all)
# trans_ab の、両商品を含むバスケットの数が全体のバスケットに占める割合
# Display the ratio of the number basket which contains both item A and B to the total number of baskets.

print('両商品を含むバスケットが全体に占める割合:{:.3f}'.format(len(trans_ab)/len(trans_all)))
print('The number of baskets which contains both item A and B to the total number of baskets:{:.3f}'.format(len(trans_ab)/len(trans_all)))

0.014という数字がでていますが、これが高いか低いかは相対的な比較になりますので、一概にはいえません。<br>
複数の商品の組み合わせで、「どれが最も一緒に売れやすいか」という比較の目安になります。<br>
一般的に支持度の小さいルールは有用性も低いことが多いことから、支持度は足切り基準として使われたりします。<br>
<br>
The result, 0.014, needs to be compared with the support of other combinations of items.<br>
If it is higher than that of other combinations, it is more likely that the combination of the items is sold at the same time.<br>
A combination of items with low support is not promissing at sales, support can be used to cut off unpromissing items.<br>

### <b>注意!!!</b>
リフト値などの計算をするときですが、StockCodeが文字列を含んでいるときといないときでちょっと書き方が違うので、注意してください。

<b>85123A、のように、文字が含まれていると、数字が入っていても文字列として解釈されるので、クオーテーションマークが要りますが、84879は数字のみで構成されているので、数字として扱われます。</b>

<b>If the StockCode includes characters, it is assumed to be a string. If not, it is assumed to be a number. In this case, you don't have to use quotation mark. Please see the examples below.</b>

〇クオーテーションマークが要る場合 / Need to use quotation mark<br>
trans_a = set(trans[trans['StockCode']=='85123A'].InvoiceNo)

〇クオーテーションマークが要らない場合 / No need to use quotation mark<br>
trans_a = set(trans[trans['StockCode']==12345].InvoiceNo)

### ルールを構成する商品に対する支持度
### Support of an item with respect to the entire transaction set
<img src="img/w3_support2.png" width="400px"><br>
支持度はルールの支持度だけではなく、ルールを構成する商品に対する支持度を計算することもあります。<br>
この支持度はリフト値を計算するときに使います。<br>
たとえば商品85123Aの支持度は、以下のように計算できます。<br>

Support can be calculated for an item with respect to the entire transaction set.<br>
This support is used when you calculate a value, lift.<br>
For example, support of item "85123A" is calculated as follows.

In [ ]:
# support{A} = A / N
# A = len(trans_a), N = len(trans_all)
print('商品85123Aを含むバスケットの全体に占める割合:{:.3f}'.format(len(trans_a)/len(trans_all)))
print('Supprt for item "85123A":{:.3f}'.format(len(trans_a)/len(trans_all)))

<a id="confidence"></a>

### 確信度 (confidence)
<img src="img/w3_confidence.png" width="400px"><br>
確信度とは、ある商品Aの購入数をベースに、その商品Aとある商品Bの組み合わせ購買がどれくらいの割合であるのかを表します。<br>
商品85123Aを購入するなら商品85099Bも購入する、というルールの確信度は、次のように計算できます。<br>
<br>
Confidence is the ratio of the number of baskets which include both item A and B to the total number of baskets which includes item A.<br>
The confidence {"85123A"->"85099B"} is calculated as follows.<br>

In [ ]:
# Confidence = (A and B) / A
# A and B = len(trans_ab)
# A = len(trans_a)
print('Confidence("85123A"->"85099B"):{:.3f}'.format(len(trans_ab)/len(trans_a)))

逆に、商品85099Bを購入するなら商品85123Aする、というルールの確信度は以下のようになります。<br>
The confidence {"85099B"->"85123A"} is calculated as follows.<br>

In [ ]:
print('Confidence("85099B"->"85123A"):{:.3f}'.format(len(trans_ab)/len(trans_b)))

確信度が高いと商品間の併売が見込めるため、クロスセル（他の商品などを併せて購入してもらうこと）をさせたい場合は確信度の高い商品の中からオファー商品を決定するなどの場面で活用されます。ただし、確信度の絶対値だけでは併売傾向の判断を誤ることがあるため、次のリフト値も合わせて見るのが普通です。<br>
<br>
When the confidence is high, it is likely that customers may purchase these two items at the same time.<br>
So it will be a good candidate as a recommended item.<br>
But the magnitude value of confidence is not always reliable, so it is better to decide a recommendation with lift value.

<a id="lift"></a>

### 9-4-3-3 リフト値 (lift)
<img src="img/w3_lift.png" width="400px"><br>
商品Aを購入するなら商品Bを購入するというアソシエーションルールにおいて、そのリフトとは、ルールの確信度（%）を商品Bの支持度（%）で割った値のことです。<br>
つまり、全体のバスケットに占める商品Bの購買率に対する、商品Aを購買したときの商品Bの購買率の比率がリフト値です。当然、リフト値が1.0よりも大きければ併売しやすい商品になりますし、1.0よりも小さければ併売しにくい商品と解釈できます。商品85123Aを購入するなら商品85099Bも購入する、というルールのリフトは以下のように求められます。<br>


In [ ]:
# Lift = (A and B) / (A * B)
# A and B = len(trans_ab)
# A = len(trans_a), B = len(trans_b), N = len(trans_all)

# 全体のバスケットに占める商品Bの購買率を計算
support_b = len(trans_b) / len(trans_all)

# 商品Aを購買したときの商品Bの購買率を計算
confidence = len(trans_ab) / len(trans_a)

# リフト値を計算
lift = confidence / support_b
print('lift:{:.3f}'.format(lift))

確信度が高い数値でもリフト値が1.0を下回る場合は、顧客への商品推奨の根拠としては不適切かもしれません。分析の目的に照らし、確信度とリフト値を組み合わせて使うなどしましょう。<br>
また、店舗エリア別、店舗タイプ別、顧客クラスター別などにルールを抽出すると、より有用性の高いルールが抽出できるかもしれないことを覚えておきましょう。<br>
<br>
Even if confidence is high, items which has lift below 1.0 may not be a good combination to be recommended.<br>
Recommendations have to be considered with the three values (support, confidence, and lift) in a comprehensive manner.<br>
You may be able to extract more useful association rules by analyzing classified data by area and type of the stores or by clustering of the customers.<br>


## MLXTEND を使うと、自動でいろいろな組み合わせのサポート、リフト、確信度を調べてくれます。


In [ ]:
# ライブラリのインストール
# ! pip install mlxtend
! conda install -c conda-forge mlxtend

In [ ]:
# MLXTEND から関連ライブラリをインポート
from mlxtend.frequent_patterns import association_rules
from mlxtend.frequent_patterns import apriori

## 商品名を列にして、InvoiceNo（請求書番号）ごとに購入した商品数を表すデータ書式に変える。

In [ ]:
basket = (trans.groupby(['InvoiceNo', 'Description'])['Quantity']
          .sum().unstack().reset_index().fillna(0)
          .set_index('InvoiceNo'))
basket.head(10)

## 購入したら1、しなかったら0、にデータを書き換える

In [ ]:
def sum_to_boolean(x):
  if x<=0:
    return 0
  else:
    return 1
basket_final = basket.applymap(sum_to_boolean)
basket_final.head(10)

## 少なくとも 2% のサポートが含まれているアイテムのセットを表示します。
少なくとも、2%のバスケットで購入されているアイテムのみを抽出します。

データサイズが大きいと、それだけいろいろな種類の商品が含まれるので、サポートは小さめに出ます。<br>
データサイズが小さいものを扱うときは、サポートの最小値も5%、6%というように少し大きめに取りましょう。<br>

In [ ]:
frequent_itemsets = apriori(basket_final, min_support = 0.02,
  use_colnames = True)
# frequent_itemsets.sort_values('support', ascending = False).head()
frequent_itemsets.sort_values('support', ascending = False)

In [ ]:
frequent_itemsets.shape

## 関数association_rulesを使って、自動で3つの指標を計算

In [ ]:
# 購買パターンをリフト値で並べ替え
a_rules = association_rules(frequent_itemsets, metric = "lift",  min_threshold = 1)
a_rules.sort_values('lift',ascending = False)

## どんな組み合わせの商品がよく売れているでしょうか??